src/backend/tcop/zenith_wal_redo.c

```diff
@@ -433,84 +434,45 @@ pprint_tag(BufferTag *tag)
  *	EOF is returned if end-of-file input is seen; time to shut down.
  * ----------------
  */
-
-/*
- * Wait until there is data in stdin. Prints a log message every 10 s whil
- * waiting.
- */
-static void
-wait_with_timeout(void)
-{
-	for (;;)
-	{
-		struct timeval timeout = {10, 0};
-		fd_set		fds;
-		int			ret;
-
-		FD_ZERO(&fds);
-		FD_SET(STDIN_FILENO, &fds);
-
-		ret = select(1, &fds, NULL, NULL, &timeout);
-		if (ret != 0)
-			break;
-		elog(DEBUG1, "still alive");
-	}
-}
-
 static int
 ReadRedoCommand(StringInfo inBuf)
 {
-	char		c;
+	char		hdr[1 + sizeof(int32)];
 	int			qtype;
 	int32		len;
-	int			nread;
-
-	/* FIXME: Use unbuffered I/O here, because the WAL redo process was getting
-	 * stuck with buffered I/O. I'm not sure why, or whether the bug was somewhere
-	 * in here or in the calling page server side.
-	 */
-	wait_with_timeout();
-	if (read(STDIN_FILENO, &c, 1) == 0)
-		return EOF;
-	qtype = c;
 
-	/*
-	 * Like in the FE/BE protocol, all messages have a length word next
-	 * after the type code; we can read the message contents independently of
-	 * the type.
-	 */
-	if (read(STDIN_FILENO, &len, 4) != 4)
+	/* Read message type and message length */
+	if (fread(hdr, 1, sizeof(hdr), stdin) != sizeof(hdr))
 	{
```

Comment:
May be we should alter default stdio buffer size? May be it is not needed because size of passed data is not expected to be very large (~BLCKSZ).